[quant] quantized GLU and HardGLU

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42443 [quant] quantized GLU and HardGLU**
* #42728 HardGLU implementation
* #42727 Factoring out common code in the GLU
* #42442 [quant] Quantization parameters in TensorIterator

Differential Revision: [D23011802](https://our.internmc.facebook.com/intern/diff/D23011802)

Commit:
[quant] quantized GLU and HardGLU

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2595,60 +2692,74 @@ void dequantize_tensor_per_channel_affine_cpu(
-REGISTER_DISPATCH(qtanh_stub, &qtanh_kernel);
-REGISTER_DISPATCH(qhardswish_stub, &qhardswish_kernel);
-REGISTER_DISPATCH(qelu_stub, &qelu_kernel);
+REGISTER_DISPATCH(
```
Sorry -- this is just sorting alphabetically

z-a-f commented on aten/src/ATen/native/quantized/cpu/quantized_ops.h:
```diff
@@ -152,35 +154,36 @@ using qnormalize_fn = void (*)(
-DECLARE_DISPATCH(qhardswish_fn, qhardswish_stub);
-DECLARE_DISPATCH(qelu_fn, qelu_stub);
-DECLARE_DISPATCH(qmaxpool_2d_fn, qmaxpool_2d_nhwc_stub);
+DECLARE_DISPATCH(activation_fn, qglu_stub);
```
Sorting here again -- was really hard to find a kernel in the list

vkuzo reviewed (commented):
makes sense, should we add a test as well?

Commit:
Update on "[quant] quantized GLU and HardGLU"




[ghstack-poisoned]

Commit:
Update on "[quant] quantized GLU and HardGLU"




[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -700,6 +700,103 @@ void qthreshold_kernel(
+      [&](scalar_t qa, scalar_t qb) -> scalar_t {
+        const auto a = at::native::dequantize_val(a_sc, a_zp, qa);
+        const auto b = at::native::dequantize_val(b_sc, b_zp, qb);
+        const auto y = a * std::min(std::max(b + 3.0f, 0.0f), 6.0f) / 6.0f;
```
nit: we can use `three` and `six`?

vkuzo reviewed (commented):
can we also add a benchmark?

Commit:
Update on "[quant] quantized GLU and HardGLU"




[ghstack-poisoned]

Commit:
Update on "[quant] quantized GLU and HardGLU"


Differential Revision: [D23011802](https://our.internmc.facebook.com/intern/diff/D23011802)

[ghstack-poisoned]

Commit:
Update on "[quant] quantized GLU and HardGLU"


Differential Revision: [D23011802](https://our.internmc.facebook.com/intern/diff/D23011802)

[ghstack-poisoned]

Commit:
Update on "[quant] quantized GLU and HardGLU"


Differential Revision: [D23011802](https://our.internmc.facebook.com/intern/diff/D23011802)

[ghstack-poisoned]
